### PR TITLE
Add CollectionsMarshal.GetValueRefOrAddDefault

### LINKED ISSUE #27062: Developers can use CollectionsMarshal ref accessors for Dictionary<TKey, TValue>
<details>
<summary>
Edited by @layomia. Original proposal by @benaadams (click to view)
</summary

They aren't "safe" operations as the backing store can change if items are added, removed, etc and the changes aren't tracked 
```csharp
namespace System.Runtime.InteropServices
{
    public partial static class CollectionsMarshal
    {
        public ref TValue ValueRef<TKey, TValue>(Dictionary<TKey, TValue> dict, TKey key);
    }
}
```
However it is useful to get a reference to modify struct `TValue` entries without a full copy and replace update.

/cc @jkotas 

/cc @stephentoub is it valid for `ConcurrentDictionary`?
</details>

---

# CollectionsMarshal ref accessors for `Dictionary<TKey, TValue>`

Attempting to obtain a value in a dictionary, and adding it if not present, is a common scenario in .NET applications. The existing mechanism for achieving this today is by using [`Dictionary<TKey, TValue>.TryGetValue(Tkey key, out TValue value)`](https://docs.microsoft.com/dotnet/api/system.collections.generic.dictionary-2.trygetvalue?view=net-5.0#System_Collections_Generic_Dictionary_2_TryGetValue__0__1__), and adding the value to the dictionary if not present. This causes duplicate hash lookups:

```cs
if (!dictionary.TryGetValue(key, out MyType myValue)) // hash lookup
{
    myValue = CreateValue();
    dictionary.Add(key, myValue); // duplicate hash lookup
}
```

Another scenario is updating `struct` values in dictionaries. The existing pattern for achieving this causes `struct` copies and duplicate hash lookups, which potentially have non-trivial performance costs for large structs:

```cs
struct LargeStruct
{
    // Other members...
    
    public int MyInt { get; set; }

    // Other members...
}

LargeStruct myValue = dictionary[key]; // hash lookup, struct copy
myValue.MyInt++;
dictionary[key] = myValue; // another hash lookup, another struct copy
```

## Motivation

- Provide a mechanism which avoids duplicate lookups when obtaining dictionary values which may not be present.
- Provide a mechanism which avoids copies and duplicate lookups when mutating `struct` dictionary values. 

## API proposal

```cs
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal
    {
        /// <summary>
        ///   Gets a reference to the value associated with the specified key, or throws a KeyNotFoundException
        /// </summary>
        public static ref TValue GetValueRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key);

        /// <summary>
        ///   Gets a reference to the value associated with the specified key, or returns Unsafe.NullRef<TValue>
        /// </summary>
        public static ref TValue TryGetValueRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key, out bool exists);

        /// <summary>
        ///   Gets a reference to the value associated with the specified key, or inserts it with value default(TValue).
        /// </summary>
        public static ref TValue GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key, out bool exists);
    }
}
```

[`CollectionsMarshal`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal?view=net-5.0) is an unsafe class that provides a set of methods to access the underlying data representations of collections.

### Updating `struct` value in dictionary: `KeyNotFoundException` thrown when key not present in dictionary

This pattern is helpful when caller wants to optimally update a `struct` in a scenario where the key being absent is an error state. Creating a value and adding it to the dictionary, if not already present, is not desired.

```cs
try
{
    ref MyType value = CollectionsMarshal.GetValueRef(dictionary, key);
    value.MyInt++;
}
catch (KeyNotFoundException exception)
{
    // Handle exception
}
```

### `Unsafe.NullRef<TValue>()` returned when key not present in dictionary

This pattern satisfies both the optimal `struct` value update and optimal "get or add" value scenarios.

```cs
ref MyType value = CollectionsMarshal.TryGetValueRef(dictionary, key, out bool exists);

if (exists)
{
    value.MyInt++;
}
else
{
    ref value = new MyType() { MyInt = 1 };
}
```

### `default(TValue)` returned when key not present in dictionary

This pattern also satisfies both the optimal `struct` value update and optimal "get or add" value scenarios. A `struct` `default` value always being instantiated may cause the `TryGetValueRef` to be preferred, depending on the perf scenario.

```cs
ref MyType value = CollectionsMarshal.GetValueRefOrDefault(dictionary, key, out bool exists);
value.MyInt++;

if (exists)
{
    // Do something if I care that the key already existed.
}
```

## Alternative design

`GetOrAdd` methods, similar to those on [`ConcurrentDictionary<TKey, TValue>`](https://docs.microsoft.com/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-5.0), were proposed in https://github.com/dotnet/runtime/issues/15059.

```cs
public class Dictionary<TKey, TValue>
{
    public TValue GetOrAdd([NotNull] TKey key, Func<TKey, TValue> valueFactory);
}

/* OR */

public class Dictionary<TKey, TValue>
{
    public TValue GetOrAdd<TState>([NotNull] TKey key, TState state, Func<TKey, TState, TValue> valueFactory);
}
```

### Upsides

- More discoverable/friendly than the relatively advanced APIs in this proposal.
- Follows precendent in the BCL i.e. `ConcurrentDictionary<TKey, TValue>`.

### Downsides

- Not pay-for-play. Since the new methods would live in `System.Private.CoreLib`, generic expansion issues due to `struct`-based generics will affect all users of `Dictionary<TKey, TValue>`, not just those that use them.
    - An argument against this point is that the methods can live in an extensions class, but that won't provide any perf benefits since they won't be able to access non-`public` members.
- Doesn't address the issue of copying when mutating large `struct` values.

### Should the `out bool exists` parameter on `TryGetValueRef` be removed? 

Since a call to `Unsafe.IsNullRef<T>(ref value)` can indicate whether the value exists in the dictionary, the second method could simply be:

```cs
public static class CollectionsMarshal
{
    /// <summary>
    ///   Gets a reference to the value associated with the specified key, or returns Unsafe.NullRef<TValue>
    /// </summary>
    public static ref TValue TryGetValueRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key);
}
```

#### Usage

```cs
ref MyType value = CollectionsMarshal.TryGetValueRef(dictionary, key);

if (!Unsafe.IsNullRef(ref value))
{
    value.MyInt++;
}
else
{
    ref value = new MyType() { MyInt = 1 };
}
```

### Any concerns about API bloat in the `CollectionsMarshal` type?

The generic expansion highlighted in the `GetOrAdd`-based alternative doesn't apply much here since the new methods will live in `CollectionsMarshal` and will be pay-for-play. However, are there concerns about bloating the type?

One answer [here](https://github.com/dotnet/runtime/issues/27062#issuecomment-776850067) states as follows:

> I'd think to the same extent as any other/similar type such as MemoryMarshal, Unsafe, etc.
>
> As it is today, we have 1 method in it and this proposal is adding 2-3 more. We only have a limited number of collection types in S.P.Corelib and a more limited set where returning a ref makes sense, so I don't think we are at risk of overloading it.

### PR DESCRIPTION
**Pull request by Sergio0694**
### Contributes to https://github.com/dotnet/runtime/issues/27062

```csharp
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal
    {
        public static ref TValue? GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key, out bool exists)
            where TKey : notnull;
    }
}
```

> **NOTE:** I've marked the returned `TValue` as an uncostrained nullable value, because if `TValue` is a reference type the returned reference will point to `null`. Value types are not affected by this, ie. there is no `Nullable<T>` involved here at all.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:684]: Ideally, this would not add any new code to Dictionary to make this niche method pay-for-play.

Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:684]: I was hoping for `Dictionary` to already have an equivalent method to just expose like it was done for `GetValueRefOrAddRef`, but unfortunately there doesn't seem to be one and I'm not really seeing a way to implement this API without either adding some code (which I did so far) or at least making some changes to `TryInsert` (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed. Not really sure how to proceed here, feedbacks are more than welcome 😄

One random idea I had to move the logic out of the `Dictionary` type was to refactor the code to be like:
```csharp
class Dictionary
{
    internal class Impl
    {
        public internal ref TValue? GetValueRefOrAddDefault(TKey key, out bool exists);
    }
}
```

So that `Impl.GetValueRefOrAddDefault` could access private members of the dictionary but without the method adding overhead to the dictionary type itself. Though this approach is adding another type, so it might not be desireable either. Another idea was to move the whole method logic to `CollectionsMarshal`, but that'd have required to make all the necessary dictionary members internal (and not private), so that seemed at the very least quite a bit ugly 😕

Thoughts?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:684]: > without either adding some code (which I did so far) or at least making some changes to TryInsert (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed.

Now you know why @benaadams didn't add this ;-)

Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:684]: It was a trap 🤣

**Comment by Sergio0694** [comment]
@jkotas Could you clarify what the overhead of this new method would be, and whether it would be a blocker for this PR? This would be a very neat feature to have, and the API itself has been approved by API review too - if it's just the current implementation that's an issue would any of the proposed approaches help? As in, moving the code to a diferent spot, etc.

Thanks! 🙂

**Comment by jkotas** [comment]
It makes every Dictionary instantiation more expensive, in particular with AOT.

**Comment by Sergio0694** [comment]
I see. If the issue is specifically with the overhead of dictionary instantiations, would moving the code to a separate type (eg. a nested type within dictionary) solve that? That would only ever be loaded when the method is actually used, right? 🤔

**Comment by jkotas** [comment]
> would moving the code to a separate type (eg. a nested type within dictionary) solve that

Yes, it would solve that.

**Comment by Sergio0694** [comment]
Test failures should be unrelated, and the changes in https://github.com/dotnet/runtime/pull/54611/commits/99523ba7ee43fb2c0f40f111387f837e668880f1 addresses Jan's concerns on additional overhead being added to `Dictionary<TKey, TValue>` instantiations (moved all the new code to a separate type). Should be good to review now 😄

**Review by GrabYourPitchforks — COMMENTED** [review]
@Sergio0694 Thanks for this! Can you also add a unit test for the randomized hash logic? I think the appropriate place to add this would be in the file https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections/tests/Generic/Dictionary/HashCollisionScenarios/OutOfBoundsRegression.cs.

There's no need to modify `RunDictionaryTest` (line 93 of that file) to run a `CollectionsMarshal` test for every possible combination; that's probably overkill. Instead, what you can do is immediately _before_ line 93, insert a manual call that looks something like:

```cs
RunCollectionTestCommon(
    () => new Dictionary<string, object>(StringComparison.Ordinal),
    (dictionary, key) => CollectionsMarshal.GetRef(dictionary, key, out _) = null,
    (dictionary, key) => dictionary.ContainsKey(key),
    dictionary => dictionary.Comparer,
    expectedInternalComparerTypeBeforeCollisionThreshold: nonRandomizedOrdinalComparerType,
    expectedPublicComparerBeforeCollisionThreshold: StringComparer.Ordinal,
    expectedInternalComparerTypeAfterCollisionThreshold: randomizedOrdinalComparerType);
```

In a nutshell, this utilizes the existing test infrastructure to generate a whole bunch of collisions, but the delegate you provide will force it to use your new `CollectionsMarshal` method instead of `Add`. This will get regression test coverage for string collision scenarios in this method.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs
Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs]: Whoops, fixed in https://github.com/dotnet/runtime/pull/54611/commits/ab97762a21634f0d7d863ca6dc656bd12d583cb2, thanks! 😄

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:684]: Since this is essentially duplicating the logic of `Dictionary.TryInsert`, could you add a comment in both methods advising maintainers to keep them in sync?

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs:307]:
```suggestion
            // The two APIs should behave the same when values already exist.
```

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs:367]: Shouldn't there also be an assertion here validating that `entry3Ref` equals `default(Struct)`?

**Review by eiriktsarpalis — APPROVED** [review]
Other than minor feedback, LGTM. Thanks!

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs
Comment (Sergio0694) [reply src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs]: Fixed as part of 5209a5bedff52a33454da401854502d6b967501a 👍

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs
Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:687]: Done in 5209a5bedff52a33454da401854502d6b967501a 🙂

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs
Comment (Sergio0694) [reply src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs:367]: Right, yeah it was present in the ref type case below but not here. Fixed in 21ab729485561a6c292a90ebc5a1c28182d3f64a, thanks! 😄

**Comment by GrabYourPitchforks** [comment]
A reminder for the future - try to avoid rebase & force-push after reviews have started. Makes it a little more complicated for us since we can't rely on the GitHub web UI and instead we need to generate incremental diffs manually.

**Comment by Sergio0694** [comment]
Oh, I'm sorry, I didn't realize that. Will definitely avoid doing that again in the future! 😅

**Review by GrabYourPitchforks — APPROVED** [review]
Thanks so much! :D

**Pull request merged**
